tracer/src/Datadog.Trace.Coverage.collector/CoverageCollector.cs

```diff
@@ -0,0 +1,253 @@
+            }
+        }
+
+        private void OnSessionEnd(object? sender, SessionEndEventArgs e)
+        {
+            _logger?.SetContext(e.Context);
+        }
+
+        private void ProcessFolder(string folder, SearchOption searchOption)
+        {
+            if (_tracerHome is null)
+            {
+                return;
+            }
+
+            var processedDirectories = new HashSet<string>();
+            var numAssemblies = 0;
+            var tracerAssemblyName = typeof(Tracer).Assembly.GetName().Name;
+
+            // Process assemblies in parallel.
+            Parallel.ForEach(
+                Directory.EnumerateFiles(folder, "*.*", searchOption),
+                file =>
+                {
+                    // Skip the Datadog.Trace assembly
+                    if (tracerAssemblyName == Path.GetFileNameWithoutExtension(file))
+                    {
+                        return;
+                    }
+
+                    var extension = Path.GetExtension(file).ToLowerInvariant();
+                    if (extension is ".dll" or ".exe")
+                    {
+                        try
+                        {
+                            var asmProcessor = new AssemblyProcessor(file, _tracerHome, _logger, _ciVisibilitySettings);
+                            asmProcessor.Process();
+
+                            lock (processedDirectories)
+                            {
+                                numAssemblies++;
+                                processedDirectories.Add(Path.GetDirectoryName(file) ?? string.Empty);
+                            }
+                        }
+                        catch (Datadog.Trace.Ci.Coverage.Exceptions.PdbNotFoundException)
+                        {
+                            // If the PDB file was not found, we skip the assembly without throwing error.
+                        }
+                        catch (BadImageFormatException)
+                        {
+                            // If the Assembly has not the correct format (eg. native dll / exe)
+                            // We skip processing the assembly.
+                        }
+                        catch (Exception ex)
+                        {
+                            _logger?.Error(ex);
+                        }
+                    }
+                });
+
+            // Add Datadog.Trace dependency to the deps.json
+            if (processedDirectories.Count > 0)
+            {
+                foreach (var directory in processedDirectories)
+                {
+                    var version = typeof(Instrumentation).Assembly.GetName().Version?.ToString();
+                    foreach (var depsJsonPath in Directory.EnumerateFiles(directory, "*.deps.json", SearchOption.TopDirectoryOnly))
+                    {
+                        var json = JObject.Parse(File.ReadAllText(depsJsonPath));
+                        var libraries = (JObject)json["libraries"];
```

Comment:
nit: this json fixing code feels like it would be easy to unit-test with some snapshot tests